skbio/stats/distance/_base.py

```diff
@@ -684,12 +707,42 @@ def _subset_to_dataframe(
         # included here so that np.hstack works in the event that either i_ids
         # or j_ids is empty.
         values = [np.array([])]
-        for i_idx in i_indices:
-            i.extend([self.ids[i_idx]] * j_length)
-            j.extend(j_labels)
+        if self._flags["CONDENSED"]:
+            diagonal_val = getattr(self, "_diagonal", 0.0)
+            n = self.shape[0]
+
+            # precompute to avoid repeated calculations
```

Comment:
This part of code looks inefficient to me. But I don't immediately have a better idea either. Dictionary is less efficient than an array of indices, if you can. But since the input indices are merely a subset of all indices, dictionary might be more justified. However, I guess there is a way that utilizes NumPy fancy indexing to accelerate this. Maybe this can be saved to a future pull request.